### ISSUE TITLE
Missing TypeScript Nodes (2)

### ISSUE THREAD

**Issue by despairblue** [prettier/prettier#1480]
# **See https://github.com/prettier/prettier/issues/1480#issuecomment-299230108 for up to date info**

There are still some nodes missing, but we're getting there :)

Missing printer implementation:

* [x] "TSAsyncKeyword" (#1483)
* [x] "TSDecorator" (#1488)
* [x] "TSExportKeyword" (#1483)
* [x] "TSNamespaceFunctionDeclaration" (#1483)
* [x] "TSProtectedKeyword" (#1483)
* [x] "TSPublicKeyword" (#1483)
* [x] "TSHeritageClause" (#1483)
* [x] "TSExpressionWithTypeArguments (#1483)

Missing in `typescript-ast-nodes.js`: 

* [x] "TSParameterProperty"
* [x] "TSTypeParameter"

There are also some nodes printed incorrectly, probably because some optional keywords aren't printed which should be easy to fix. And there are errors where we're trying to access properties on undefined, those are probably because TypeScript types are missing in `typescript-ast-nodes.js` or because we expect optional stuff to just be there. 

The raw output of running prettier of the TypeScript test suite can be found [here](https://gist.githubusercontent.com/despairblue/79e090660fbb4093b35ab3b24c617dfb/raw/b16276ff18d5140cd2cc8dffe4c43669737f63bf/errors) (careful, it's big).

If anyone wants to help with this drop a line here, so we don't duplicate the effort.  

see #1422 #13 #1306

/cc @azz

**Comment by azz** [comment]
I'l taking a look at TSExportKeywotd at the moment.

**Comment by azz** [comment]
I also added printer for TSHeritageClause and TSExpressionWithTypeArguments.

**Comment by udnisap** [comment]
this is something im interested too. I can work on few as well

**Comment by vjeux** [comment]
As I'm reading through answers from my tweet ( https://twitter.com/Vjeux/status/859444410869727232 ), a lot of people are eagerly waiting for TypeScript support to push it through their codebase at work. You are working on something really impactful!

**Comment by vjeux** [comment]
After #1483, down to 
- 526: prettier(input) !== prettier(prettier(input))
- 502: ast(input) !== ast(prettier(input))

So, this divided the number of broken tests by 2!! Keep it up :)

**Comment by azz** [comment]
@vjeux Think we should focus on all of those except `prettier(input) !== prettier(prettier(input))`, and tackle that last?

**Comment by vjeux** [comment]
@azz in my experience, that category contains real issues as well. But sure, the ordering doesn't really matter :) Bringing down the others are also likely going to improve the first as well.

**Comment by bakkot** [comment]
Worth noting that typescript-eslint-parser had a [breaking change](https://github.com/eslint/typescript-eslint-parser/pull/241) recently, which will presumably need fixing here.

**Comment by azz** [comment]
@bakkot  Good point, we're pinned at `a294af`, which is a couple of commit behind that change.

**Comment by despairblue** [comment]
Here is a list of unique errors that still happen:

* [ ] Error: Comment "*param help" was not printed. Please report this error!
* [ ] Error: Comment "*type" was not printed. Please report this error!
* [ ] Error: Comment "closing comment" was not printed. Please report this error!
* [ ] Error: Comment "error -- declarations must be adjacent" was not printed. Please report this error!
* [ ] Error: Comment "error" was not printed. Please report this error!
* [ ] Error: Comment "error, not same type" was not printed. Please report this error!
* [ ] Error: Comment "inherits bar. But BB is abstract, so this is OK." was not printed. Please report this error!
* [ ] Error: Comment location overlaps with node location
* [ ] Error: Value undefined is not a valid document
* [ ] Error: did not recognize object of type "TSParameterProperty"
* [ ] Error: did not recognize object of type "TSTypeParameter"
* [ ] Error: unknown type: "TSAbstractKeyword"
* [ ] Error: unknown type: "TSConstKeyword"
* [ ] TypeError: Cannot read property 'expression' of undefined
* [ ] TypeError: Cannot read property 'replace' of undefined
* [ ] TypeError: Cannot read property 'type' of null
* [ ] TypeError: Cannot read property 'type' of undefined

And [here is the whole output](https://gist.githubusercontent.com/despairblue/79e090660fbb4093b35ab3b24c617dfb/raw/99e47ad1e34c32ddd91c15315f3c4fb92fa37d65/errors) (careful big file).

So we're exceptionally close to not crashing on any valid TypeScript :smile:.

Whoever wants to try to fix any error, drop a line. I'll start with `TSConstKeyword`. Happy hacking!

**Comment by despairblue** [comment]
Working on `Value undefined is not a valid document` now. `typescript-eslint-parser` does not add the `instanceof` as an operator to `LogicalExpression`.

**Comment by vjeux** [comment]
I'm working on `TypeError: Cannot read property 'type' of undefined`, this is about printing decorators

**Comment by bakkot** [comment]
> typescript-eslint-parser does not add the instanceof as an operator to LogicalExpression

Sounds like a bug you should file (or PR) over there.

**Comment by despairblue** [comment]
@bakkot I'm trying to fix this and open a PR :)

**Comment by vjeux** [comment]
I just opened https://github.com/eslint/typescript-eslint-parser/issues/250 on typescript-eslint-parser for decorators being inconsistently outputted and #1509 as a workaround on prettier, but I'd rather not ship it and get it fixed upstream.

**Comment by JamesHenry** [comment]
`instanceof` fix merged, thanks for that! Literally no idea why I didn't PR that working branch before 😄

**Comment by JamesHenry** [comment]
Just FYI I am waiting to hear back from @Pajn on https://github.com/eslint/typescript-eslint-parser/pull/196

I am assuming that is currently blocking a related prettier PR, although one is not mentioned specifically

**Comment by vjeux** [comment]
I've been asking on Twitter for some real codebases to try it on and got great results ( https://twitter.com/Vjeux/status/860702716309364737 )
- https://github.com/palantir/tslint
- https://github.com/apollographql/apollo-client
- https://github.com/Microsoft/vscode
- https://github.com/infernojs/inferno
- https://github.com/ceolter/ag-grid

We should ensure to have no ast differences on all those projects before getting people to use it :)

**Comment by vjeux** [comment]
For apollo-client: over 91 files, 16 don't parse after prettier, 13 generate a different ast and 5 are not stable. It's not as bad as I expected :)

https://gist.github.com/vjeux/8088979c0f53922c9948e8d8c693a0ae

**Comment by azz** [comment]
@vjeux Looking good. Is there a way to print diffs ignoring whitespace and double quote/single quote changes? There's a lot of noise in the debug output.

Opened a couple more PRs to fix some of the `ast(input) !== ast(prettier(input))` issues.

**Comment by vjeux** [comment]
@azz, wanna go to https://gitter.im/jlongster/prettier so we can chat in real time?

**Comment by vjeux** [comment]
I created 3 issues found from apollo-client `ast(input) !== ast(prettier(input))`. I believe that once we fix the three, we'll be good to go for this class of bug on this project.

https://github.com/prettier/prettier/issues?q=is%3Aissue+is%3Aopen+label%3ATypeScript

**Comment by azz** [comment]
[Latest full TypeScript debug check output](https://gist.githubusercontent.com/azz/03beb5fe6fc081e40de1ec1a0c6a969e/raw/8da0b731867f255697b2b87c76694d8d20bec6a1/..-prettier.results.txt) (large!)

**Comment by vjeux** [comment]
the current output of syntax error is really bad. For babylon and flow we feed it through a pretty printed that shows the snippet of code. Should be easy to setup and help a lot :)

**Comment by vjeux** [comment]
And would be sweet to print the various counts at the end of debug check, this way it's easier to track progress

**Comment by azz** [comment]
The issue above seems to causing most of the errors in `TypeScript/tests/cases/fourslash`.

**Comment by azz** [comment]
There are a cases that are simply invalid TypeScript (at least semantically) that I'm not sure whether to raise with `typescript-eslint-parser`. One example is this:

```
../TypeScript/tests/cases/compiler/modifierOnParameter1.ts
ast(input) !== ast(prettier(input))
Index: 
===================================================================
--- 	
+++ 	
@@ -22,11 +22,6 @@
               "params": [
                 {
-                  "type": "TSParameterProperty",
-                  "accessibility": null,
-                  "isReadonly": false,
-                  "parameter": {
-                    "type": "Identifier",
-                    "name": "p"
-                  },
+                  "type": "Identifier",
+                  "name": "p",
                   "decorators": []
                 }

Index: 
===================================================================
--- 	
+++ 	
@@ -1,3 +1,3 @@
 class C {
-   constructor(declare p) { }
-}
\ No newline at end of file
+  constructor(p) {}
+}
```

**Comment by azz** [comment]
TODO:

 * [ ] `questionToken` is not printed:
    ```ts
    - [Symbol.iterator]: { x };
    + [Symbol.iterator]?: { x };
    ```

 * [ ] `readonlyToken` is not printed in mapped types:
    ```ts
    -var v03: { readonly [P in keyof T]: T[P] };
    +var v03: { [P in keyof T]: T[P] };
    ```

 * [ ] Do not put `[]` around `Literal` nodes on LHS in interface.
    ```ts
    interface I {
    -    [Symbol.iterator]: number;
    -    [s: symbol]: string;
    -    "__@iterator": string;
    +  [Symbol.iterator]: number,
    +  [s: symbol]: string,
    +  ["__@iterator"]: string
    }
    ```

 * [ ] Print parens around arrow functions in this case:
    ```ts
    -var x = (v => v) as (x: number) => string;
    +var x = v => v as (x: number) => string;
    ```
 *  [ ] Extraneous newline on second print. Also occurs with functions:
 
    ```ts
       export var m2_a1: number;

    +
       export class m2_c1 {
         m2_c1_p1: number;
    ```

 * [ ] Unstable re-print:
    ```ts
    class C5 {
      async f() {
    -    const x = yield* (async function*() {
    -      yield 1;
    -    })();
    +    const x =
    +      yield *
    +      (async function*() {
    +        yield 1;
    +      })();
      }
    }
    ```

**Comment by JamesHenry** [comment]
Will have some more feedback on printing soon (am working on parser stuff today), but here is a critical one I noticed after a quick test:

```ts
class Foo {
   constructor(
      public bar: string,
      private baz: number,
   ) {}
}
```

I noticed the accessibility modifiers (public, private) are currently not being printed for constructor parameters.

**Comment by azz** [comment]
@JamesHenry I can't repro that issue:

```bash
$ echo '
> class Foo {
>    constructor(
>       public bar: string,
>       private baz: number,
>    ) {}
> }' | prettier --parser typescript
class Foo {
  constructor(public bar: string, private baz: number) {}
}
```

**Comment by JamesHenry** [comment]
@azz Interesting, once I have finished my parser stuff, I will do a proper test on the full source

**Comment by Igorbek** [comment]
Update on `Error: Comment "*type" was not printed. Please report this error!`

This happens because comment cannot be attached to `TypeParameter` node as a leading comment, because `typescript-eslint-parser` incorrectly calculates location of such nodes by including leading space, including comments. So that such comment are found to be enclosed by TypeParameter node.

Issue filed eslint/typescript-eslint-parser#260
PR created eslint/typescript-eslint-parser#259

**Comment by JamesHenry** [comment]
Thanks, @Igorbek! That's merged now

**Comment by vjeux** [comment]
apollo-client output is now much smaller! Awesome work everyone :) https://gist.github.com/vjeux/b96de2e8f196e558b8c9dd0feb82430b

**Comment by JamesHenry** [comment]
Just FYI to all you wonderful people working on TS support in prettier, we have merged in some major PRs to `typescript-eslint-parser` (which powers the prettier TS parser) today.

Of particular note:

1. The `typescript-eslint-parser` codebase itself has been refactored (https://github.com/eslint/typescript-eslint-parser/pull/261), and whilst this does not change any ASTs, it will affect any of you that have forks of the project.

2. @Pajn's PR to normalize type parameters (https://github.com/eslint/typescript-eslint-parser/pull/196), this is a breaking change for relevant ASTs and so will affect (hopefully greatly improve) relevant prettier work.

There is a chance @Pajn has some relevant prettier WIP, but I know that he has been busy moving cities, so not sure of the status.

**Comment by vjeux** [comment]
If you are interested, how master looks on inferno, pretty damn close :) https://github.com/infernojs/inferno/pull/1069

**Comment by azz** [comment]
This change leads to `ast(input) !== ast(prettier(input))`, but I know it is intentional. Do I leave it as is, @vjeux?

```ts
 class C {
-    "foo"();
-    "bar"() { }
+  foo();
+  bar() {}
```

**Comment by vjeux** [comment]
@azz: if it has the same semantics, then it's fine. You can remove the warning deleting those nodes from the ast when you know it's intentional: https://github.com/prettier/prettier/blob/master/bin/prettier.js#L142-L203

**Comment by azz** [comment]
[Latest error log (after 1544)](https://gist.github.com/azz/263dbfd8a18549cef23670917e6a607b) (fits in a gist now!)

Just printing errors.
```
(find ../TypeScript/tests/ -name '*.ts' | grep -v "'" | xargs ./bin/prettier.js --parser typescript --debug-check) 2>&1 | egrep -v "^\.\.[^:]*$" | tee ../prettier.results.txt
```

**Comment by beeman** [comment]
Guys, I just wanted to let you know that I highly appreciate that you're building support for TypeScript in prettier! Looking forward to start using it across all my projects 👍

**Comment by azz** [comment]
Does someone want to take a look at the `+`/`-` operators?

(changed plus/minus from the diff to `was`/`now` to avoid confusion.)

// ok
```ts
was (1)++;
now 1++;
was (1)--;
now 1--;
 was ++(1);
now ++1;
was --(1);
now --1;

// bad 
was (1 + 2)++;
now 1 + 2++;
was (1 + 2)--;
now 1 + 2--;
 was ++(1 + 2);
now ++1 + 2;
was --(1 + 2);
now --1 + 2;
 was (x + x)++;
now x + x++;
was (x + x)--;
now x + x--;
 was ++(x + x);
now ++x + x;
was --(x + x);
now --x + x;
```

**Comment by bakkot** [comment]
@azz, none of those are syntactically valid JS, so I wouldn't worry too much about it.

**Comment by despairblue** [comment]
I started integrating `eslint-parser-typescript`'s master branch and fixing what it breaks, just so you know.

**Comment by JamesHenry** [comment]
Thanks, @despairblue!

We are trying to get in as many necessary breaking changes as possible before we declare prettier TS support "stable".

Another tiny one (in terms of the changes required in prettier) coming:
https://github.com/eslint/typescript-eslint-parser/pull/219

We will need to change the parser config to use `comment: true`, rather `attachComment: true` as it currently is.

**Comment by vjeux** [comment]
@despairblue if there are too much changes, feel free to commit the bump and have travis fail for a while until we fix all them in parallel. Doesn't have to be blocked by having done all of them. (I tried for half an hour and there was a bunch of work needed)

**Comment by JamesHenry** [comment]
I should be able to help out with some on Thursday too, as I have the day off 😄

**Comment by despairblue** [comment]
@vjeux Will do. I started on early on Monday but personal life and work took too much time to finish it up. I'll clean up my working directory tomorrow and open a PR with what I have, thank you for all the support :)

**Comment by azz** [comment]
@despairblue What's the status of the `typescript-eslint-parser` upgrade? Noticed you merged a few PRs.

I should be able to jump in over the next couple of days.

**Comment by despairblue** [comment]
@azz Only to tests are left

* [ ] `tests/typescript/conformance/types/abstractKeyword/jsfmt.spec.js`
* [ ] `tests/typescript/conformance/internalModules/importDeclarations/jsfmt.spec.js`

That needs to be fixed in typescript-eslint-parser https://github.com/eslint/typescript-eslint-parser/issues/269

There is also the problem that all comments in a TSInterfaceBody get attached to the TSInterfaceDeclaration, which makes it hard to print them in the correct place without using their location data traversing the TSInterfaceBody and attaching them to the correct nodes.

To elaborate:

> Mhh, I find it weird that for interfaces all comments are always attached to the `TSInterfaceDeclaration`, whereas for classes and objects the comments are attached to the following node:
> 
> ```ts
> interface I {
>   a: string;
>   // this comment is attached to the `TSInterfaceDeclaration` node which is right under `Program`
>   b: string;
> }
> ```
> 
> ```ts
> class C {
>   public a: string;
>   // this comment is attached to the following `ClassProperty` node
>   public b: string;
> }
> ```
> 
> ```ts
> const c = {
>   a: 1,
>   // this comment is attached to the following `Property` node
>   b: 2,
> }
> ```
> 
> To me it seems as if the printer decides where to put the comments. And for that is uses mostly the logic to print the comment before the node it is attached to. Seems in the case of `TSInterfaceDeclaration` we need to read the actual location from the comment node and use that to find out where in the `TSInterfaceDeclaration` the comment was.

**Comment by despairblue** [comment]
@azz I'm in the process of fixing the abstract and export issue (https://github.com/eslint/typescript-eslint-parser/pull/270), but the comments are a tougher problem.

**Comment by azz** [comment]
Added a script in #1606 to help whittle down the remaining failed tests from the TypeScript repo.

```
$ node scripts/run-external-typescript-tests.js
```

It groups error by type (e.g. ` Error: Value undefined is not a valid document`), and writes one file per type of error. This means you can tackle all of the comment printing issues just by looking at one file.

<details>
<summary>Latest output</summary>

```
13492 good, 1754 bad
13492/15246 files printed without errors.

624     ast(input) !== ast(prettier(input))
312     Error: Value undefined is not a valid document
110     Error: Comment "<omitted>" was not printed. Please report this error!
90      SyntaxError: ';' expected.
88      SyntaxError: Expression expected.
83      TypeError: Cannot read property 'type' of undefined
55      prettier(input) !== prettier(prettier(input))
46      SyntaxError: Invalid character.
44      SyntaxError: ',' expected.
35      SyntaxError: Identifier expected.
23      SyntaxError: Declaration or statement expected.
21      SyntaxError: '(' expected.
19      SyntaxError: Type expected.
17      SyntaxError: ':' expected.
16      SyntaxError: Property or signature expected.
16      SyntaxError: Hexadecimal digit expected.
12      SyntaxError: '=' expected.
12      SyntaxError: Unterminated template literal.
10      SyntaxError: Parameter declaration expected.
8       SyntaxError: ')' expected.
8       SyntaxError: An extended Unicode escape value must be between 0x0 and 0x10FFFF inclusive.
7       SyntaxError: 'super' must be followed by an argument list or member access.
7       SyntaxError: Unexpected token. A constructor, method, accessor, or property was expected.
6       SyntaxError: '{' expected.
6       SyntaxError: '}' expected.
6       TypeError: Cannot read property 'some' of undefined
6       SyntaxError: Property assignment expected.
5       SyntaxError: '{' or ';' expected.
4       SyntaxError: Unterminated Unicode escape sequence.
4       SyntaxError: Digit expected.
4       SyntaxError: Unterminated string literal.
3       SyntaxError: '=>' expected.
3       SyntaxError: Argument expression expected.
3       SyntaxError: Unterminated regular expression literal.
3       SyntaxError: 'as' expected.
3       SyntaxError: ']' expected.
3       Error: Comment location overlaps with node location
3       SyntaxError: An unary expression with the '-' operator is not allowed in the left-hand side of an exponentiation expression. Consider enclosing the expression in parentheses.
3       SyntaxError: Merge conflict marker encountered.
2       SyntaxError: Array element destructuring pattern expected.
2       SyntaxError: '>' expected.
2       SyntaxError: Enum member expected.
2       SyntaxError: Declaration expected.
2       SyntaxError: Variable declaration expected.
2       SyntaxError: Expression or comma expected.
2       SyntaxError: '*/' expected.
1       SyntaxError: An unary expression with the 'typeof' operator is not allowed in the left-hand side of an exponentiation expression. Consider enclosing the expression in parentheses.
1       SyntaxError: Binary digit expected.
1       SyntaxError: An unary expression with the 'delete' operator is not allowed in the left-hand side of an exponentiation expression. Consider enclosing the expression in parentheses.
1       SyntaxError: JSX element 'number' has no corresponding closing tag.
1       SyntaxError: Invalid 'reference' directive syntax.
1       SyntaxError: Type parameter declaration expected.
1       SyntaxError: 'case' or 'default' expected.
1       SyntaxError: Cannot read property 'pos' of undefined (undefined:undefined)
1       SyntaxError: JSX element 'T' has no corresponding closing tag.
1       SyntaxError: Unexpected end of text.
1       SyntaxError: 'of' expected.
1       SyntaxError: An AMD module cannot have multiple name assignments.
```
</details>

**Comment by vjeux** [comment]
New issue to follow along: https://github.com/prettier/prettier/issues/1643

**Comment by mwq27** [comment]
With prettier 1.5.3, i'm getting the error for `TSQualifiedName`.
```
Error: unknown type: "TSQualifiedName"
```
